### PR TITLE
Move part of AutoHOOT.jl to ITensorAutoHOOT

### PR DESCRIPTION
**Pull request by LinjianMa**
Also Get CI working. 

Note that since AutoHOOT.jl depends on the Python library, currently it cannot be registered to be a general library, thus the previous CI workflow doesn't work (julia-buildpkg and julia-runtest will fail). I rewrite most parts in CI to get it working.

I also removed the docs workflow, can add that back later.

### CODE REVIEW

**Comment by mtfishman** [comment]
Overall looks good.

My main suggestions would be to:

1. Make use of #5 once that is merged and remove the lower level `@adjoint` definitions in favor of those definitions.
2. Change over the higher level `@adjoint` definitions like the adjoint of `vjps` with overloads of `ChainRulesCore.rrule`, and replace the `Zygote` dependency with a `ChainRulesCore` dependency. I don't think you are actually using Zygote for anything in this module besides defining rules, right? It seems like Zygote is only actually used in the tests.

File: src/ITensorAutoHOOT/utils.jl [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+    return 'A' + i - 26
+  end
+  return Char(i + 140)
+end
+
+"""
+Retrieve the key from the dictionary that maps AutoHOOT nodes
+to ITensor tensors. Returns Nothing if key not exists.
+"""
+function retrieve_key(dict, value)
```
Comment (mtfishman) [inline src/ITensorAutoHOOT/utils.jl:18]: Note that you can use `findfirst` for this:
```julia
julia> findfirst(==(2), Dict(:a => 1, :b => 2))
:b

julia> findfirst(==(3), Dict(:a => 1, :b => 2)) # returns nothing

```

File: src/ITensorAutoHOOT/utils.jl [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+Note: this function generates a network with NEW indices
+Parameters
+----------
+outnode: AutoHOOT einsum node
+node_dict: A dictionary mapping AutoHOOT node to ITensor tensor
+Returns
+-------
+A list of ITensor tensors
+"""
+function generate_network(out_node, node_dict)
```
Comment (mtfishman) [inline src/ITensorAutoHOOT/utils.jl:111]: Is this function used anywhere?

File: src/ITensorAutoHOOT/zygote.jl [hunk 3]
```diff
@@ -0,0 +1,125 @@
+using Zygote, AutoHOOT
+using Zygote: @adjoint
+using ITensors: setinds
+
+const ad = AutoHOOT.autodiff
+const go = AutoHOOT.graphops
+
+# represent a sum of tensor networks
+mutable struct NetworkSum
```
Comment (mtfishman) [inline src/ITensorAutoHOOT/zygote.jl:9]: Does this need to be mutable?

File: src/ITensorAutoHOOT/zygote.jl [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+NetworkSum() = NetworkSum([])
+
+function Executor(networks::Array)
+  nodes, node_dict = generate_einsum_expr(networks)
+  # TODO: add caching here
+  net_sums = [NetworkSum([go.generate_optimal_tree(n)]) for n in nodes]
+  return Executor(net_sums, node_dict)
+end
+
+@adjoint Executor(networks::Array) = Executor(networks), dv -> (nothing,)
```
Comment (mtfishman) [inline src/ITensorAutoHOOT/zygote.jl:27]: Note that you can use the `@non_differentiable` macro defined in `ChainRulesCore` (https://juliadiff.org/ChainRulesCore.jl/stable/writing_good_rules.html#[@non_differentiable](@ref)).

File: src/ITensorAutoHOOT/utils.jl (hunk 1 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/utils.jl:18]: cool, will update

File: src/ITensorAutoHOOT/utils.jl (hunk 2 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/utils.jl:111]: no it's kind of a deprecated API, I will remove this

File: src/ITensorAutoHOOT/zygote.jl (hunk 3 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/zygote.jl:9]: I have some functions that update the NetworkSum.nodes, for example see line 36

Comment (mtfishman) [reply src/ITensorAutoHOOT/zygote.jl:9]: On line 36 you are only mutating the `Array` storage of the `NetworkSum` object, not the `NetworkSum` object itself. So for example it is the difference between doing:
```julia
net_sum.nodes = push!(copy(net_sum.nodes), node)
```
and
```julia
push!(net_sum.nodes, node)
```
The second one (which is what you did) does not require making the `NetworkSum` type mutable.

In general I find it is best to make things immutable unless there is a good reason, and especially if a struct already contains a mutable data structure (like in this case) there usually isn't a reason to make it mutable.

File: src/ITensorChainRules/ITensorChainRules.jl [hunk 5]
```diff
@@ -190,22 +190,24 @@ function ChainRulesCore.rrule(::typeof(itensor), x::Array, a...)
...
-    # TODO: define `Array(::ITensor)` directly
-    x̄ = Array(ȳ, inds(ȳ)...)
-    ā = broadcast(_ -> NoTangent(), a)
-    return (NoTangent(), x̄, ā...)
-  end
-  return y, ITensor_pullback
-end
+# TODO: this function is triggered in the HVP kernel and triggered error,
+# although I'm still not clear about the reason. Can debug later.
+# function ChainRulesCore.rrule(::typeof(ITensor), x::Array, a...)
```
Comment (mtfishman) [inline src/ITensorChainRules/ITensorChainRules.jl:201]: So you see an error when this` rrule` is defined, but it fixes it to remove this definition?

Comment (LinjianMa) [reply src/ITensorChainRules/ITensorChainRules.jl:201]: yes

Comment (mtfishman) [reply src/ITensorChainRules/ITensorChainRules.jl:201]: Reminder to try to add this back with the signature:
```julia
ChainRulesCore.rrule(::typeof(ITensor), x::Array{<:Number}, a...)
```
or more specifically:
```julia
ChainRulesCore.rrule(::typeof(ITensor), x::Array{<:Number}, a::Index...)
```

Comment (LinjianMa) [reply src/ITensorChainRules/ITensorChainRules.jl]: done

File: src/ITensorAutoHOOT/utils.jl (hunk 1 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/utils.jl]: Done

File: src/ITensorAutoHOOT/utils.jl (hunk 2 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/utils.jl]: Done

File: src/ITensorAutoHOOT/zygote.jl (hunk 3 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/zygote.jl]: Done

File: src/ITensorAutoHOOT/zygote.jl (hunk 4 above)
Comment (LinjianMa) [reply src/ITensorAutoHOOT/zygote.jl]: Done

**Comment by LinjianMa** [comment]
@mtfishman please let me know if you have other comments on this PR

File: test/ITensorAutoHOOT/runtests.jl [hunk 6]
```diff
@@ -1,4 +1,205 @@
...
+  B = randomITensor(j, k)
+  C = randomITensor(k, l)
+
+  out = A * B * C
+  network = [A, C, B]
+
+  nodes, dict = itensorah.generate_einsum_expr([network])
+  network = itensorah.extract_network(nodes[1], dict)
+  out2 = network[1] * network[2] * network[3]
+  @test isapprox(storage(out), storage(out2))
```
Comment (mtfishman) [inline test/ITensorAutoHOOT/runtests.jl:25]: You should be able to use `isapprox` directly on the ITensors.

Comparing the storage may fail if the tensors are permutations of each other.

File: test/ITensorAutoHOOT/runtests.jl [hunk 7]
```diff
@@ -1,4 +1,205 @@
...
+
+  out = A * B * C * D * E
+  network = [A, B, C, D, E]
+
+  nodes, dict = itensorah.generate_einsum_expr([network])
+  node = go.generate_optimal_tree(nodes[1])
+  out_list = itensorah.compute_graph([node], dict)
+  out2 = out_list[1]
+
+  @test isapprox(storage(out), storage(out2))
```
Comment (mtfishman) [inline test/ITensorAutoHOOT/runtests.jl:49]: Same comment as above.

File: test/ITensorAutoHOOT/runtests.jl [hunk 8]
```diff
@@ -1,4 +1,205 @@
...
+
+  gradA_direct = B * C * D * E
+  gradB_direct = A * C * D * E
+
+  networks = itensorah.gradients([A, B, C, D, E], [A, B])
+  gradA = contract(networks[1])
+  gradB = contract(networks[2])
+
+  @test isapprox(norm(gradA_direct), norm(gradA))
+  @test isapprox(norm(gradB_direct), norm(gradB))
```
Comment (mtfishman) [inline test/ITensorAutoHOOT/runtests.jl:94]: Seems better to compare the actually tensors, not just the norm.

File: test/ITensorAutoHOOT/runtests.jl [hunk 9]
```diff
@@ -1,4 +1,205 @@
...
+  B = randomITensor(j, k)
+  C = randomITensor(k, i)
+
+  function network(A)
+    tensor_network = [A, B, C]
+    out = itensorah.batch_tensor_contraction([tensor_network], A)
+    return sum(out)[]
+  end
+  grad_A = gradient(network, A)
+  @test isapprox(norm(grad_A), norm(B * C))
```
Comment (mtfishman) [inline test/ITensorAutoHOOT/runtests.jl:111]: Same comment as above.

File: test/ITensorAutoHOOT/runtests.jl [hunk 10]
```diff
@@ -1,4 +1,205 @@
...
+  l = Index(2, "l")
+  A = randomITensor(i, j)
+  B = randomITensor(j, i)
+  C = randomITensor(j, k)
+  D = randomITensor(k, l)
+  E = randomITensor(l, i)
+
+  function inner(A)
+    networks = [[A, B], [A, C, D, E]]
+    contract = itensorah.batch_tensor_contraction(networks, A, B, C, D, E)
```
Comment (mtfishman) [inline test/ITensorAutoHOOT/runtests.jl:157]: What is the purpose of passing the list of ITensors to this function? Couldn't you extract them from the `networks` input?

Comment (LinjianMa) [reply test/ITensorAutoHOOT/runtests.jl:157]: It is used to let Zygote know which variables to take the derivative of - in the pullback function we won't take the derivative of the network but will do for other variables.

The networks argument is not the variable of the function, it is just a graph showing what kind of einusms we want to calculate.

Comment (mtfishman) [reply test/ITensorAutoHOOT/runtests.jl:157]: I see, thanks. I see now that in the forward pass those trailing ITensors are just ignored (line 81).

Comment (mtfishman) [reply test/ITensorAutoHOOT/runtests.jl:157]: I guess it is a confusing interface if you were just thinking about using this for a forward pass, but in that case you just wouldn't pass any trailing ITensor arguments, just the `networks` input.

Comment (LinjianMa) [reply test/ITensorAutoHOOT/runtests.jl:157]: that's right

File: test/ITensorAutoHOOT/runtests.jl (hunk 7 above)
Comment (LinjianMa) [reply test/ITensorAutoHOOT/runtests.jl]: Done

File: test/ITensorAutoHOOT/runtests.jl (hunk 8 above)
Comment (LinjianMa) [reply test/ITensorAutoHOOT/runtests.jl]: Done

**Comment by mtfishman** [comment]
Looks good to me, feel free to merge.

**Review by mtfishman — APPROVED** [review]

**Pull request merged**